src/layers/python.rs

```diff
@@ -0,0 +1,582 @@
+        context: &BuildContext<Self::Buildpack>,
+        layer_data: &LayerData<Self::Metadata>,
+    ) -> Result<ExistingLayerStrategy, <Self::Buildpack as Buildpack>::Error> {
+        let cached_metadata = &layer_data.content_metadata.metadata;
+        let new_metadata = self.generate_layer_metadata(&context.stack_id);
+
+        if let Some(reason) = cache_invalidation_reason(cached_metadata, &new_metadata) {
+            log_info(format!("Discarding cache {reason}"));
+            Ok(ExistingLayerStrategy::Recreate)
+        } else {
+            log_info(format!(
+                "Using cached Python {}",
+                cached_metadata.python_version
+            ));
+            let PackagingToolVersions {
+                pip_version,
+                setuptools_version,
+                wheel_version,
+            } = &cached_metadata.packaging_tool_versions;
+            log_info(format!(
+                "Using cached pip {pip_version}, setuptools {setuptools_version} and wheel {wheel_version}"
+            ));
+            Ok(ExistingLayerStrategy::Keep)
+        }
+    }
+}
+
+impl<'a> PythonLayer<'a> {
+    fn generate_layer_metadata(&self, stack_id: &StackId) -> PythonLayerMetadata {
+        PythonLayerMetadata {
+            stack: stack_id.clone(),
+            python_version: self.python_version.to_string(),
+            packaging_tool_versions: self.packaging_tool_versions.clone(),
+        }
+    }
+}
+
+/// Metadata stored in the generated layer that allows future builds to determine whether
+/// the cached layer needs to be invalidated or not.
+#[derive(Clone, Deserialize, PartialEq, Serialize)]
+pub(crate) struct PythonLayerMetadata {
+    stack: StackId,
+    python_version: String,
+    packaging_tool_versions: PackagingToolVersions,
+}
+
+/// Compare cached layer metadata to the new layer metadata to determine if the cache should
+/// be invalidated, and if so, for what reason.
+fn cache_invalidation_reason(
+    cached_metadata: &PythonLayerMetadata,
+    new_metadata: &PythonLayerMetadata,
+) -> Option<String> {
+    // By destructuring here we ensure that if any additional fields are added to the layer
+    // metadata in the future, it forces them to be used as part of cache invalidation,
+    // otherwise Clippy would report unused variable errors.
+    let PythonLayerMetadata {
+        stack: cached_stack,
+        python_version: cached_python_version,
+        packaging_tool_versions:
+            PackagingToolVersions {
+                pip_version: cached_pip_version,
+                setuptools_version: cached_setuptools_version,
+                wheel_version: cached_wheel_version,
+            },
+    } = cached_metadata;
+
+    let PythonLayerMetadata {
+        stack,
+        python_version,
+        packaging_tool_versions:
+            PackagingToolVersions {
+                pip_version,
+                setuptools_version,
+                wheel_version,
+            },
+    } = new_metadata;
```

Comment:
I love the destructuring pattern to ensure all values are used. Style wise I don't think you need it twice though. You're guaranteed to have the same fields in each. You could:

```rust
    let PythonLayerMetadata {
        stack,
        python_version,
        packaging_tool_versions:
            PackagingToolVersions {
                pip_version,
                setuptools_version,
                wheel_version,
            },
    } = new_metadata;

    if stack != cached.stack {
        reasons.push(format!(
```

It would save some lines.

Reply:
Ahh, true. Now that you mention it. I ran into the same issue, but decided to use individual let statements inside of my branch.